cirq/circuits/circuit.py

```diff
@@ -68,74 +73,159 @@ class Circuit:
-        batch_insert
-        batch_remove
-        batch_insert_into
-        insert_at_frontier
+    @abc.abstractmethod
+    def __init__(self, *contents: 'cirq.OP_TREE',
+                 strategy: 'cirq.InsertStrategy', device: 'cirq.Device'):
+        """Initializes a circuit.
 
-    Circuits can also be iterated over,
-        for moment in circuit:
-            ...
-    and sliced,
-        circuit[1:3] is a new Circuit made up of two moments, the first being
-            circuit[1] and the second being circuit[2];
-        circuit[:, qubit] is a new Circuit with the same moments, but with only
-            those operations which act on the given Qubit;
-        circuit[:, qubits], where 'qubits' is list of Qubits, is a new Circuit
-            with the same moments, but only with those operations which touch
-            any of the given qubits;
-        circuit[1:3, qubit] is equivalent to circuit[1:3][:, qubit];
-        circuit[1:3, qubits] is equivalent to circuit[1:3][:, qubits];
-    and concatenated,
-        circuit1 + circuit2 is a new Circuit made up of the moments in circuit1
-            followed by the moments in circuit2;
-    and multiplied by an integer,
-        circuit * k is a new Circuit made up of the moments in circuit repeated
-            k times.
-    and mutated,
-        circuit[1:7] = [Moment(...)]
-    """
+        Args:
+            contents: The initial list of moments and operations defining the
+                circuit. You can also pass in operations, lists of operations,
+                or generally anything meeting the `cirq.OP_TREE` contract.
+                Non-moment entries will be inserted according to the specified
+                insertion strategy.
+            strategy: When initializing the circuit with operations and moments
+                from `contents`, this determines how the operations are packed
+                together. This option does not affect later insertions into the
+                circuit.
+            device: Hardware that the circuit should be able to run on.
+        """
+
+    @property
+    @abc.abstractmethod
+    def moments(self) -> Sequence['cirq.Moment']:
+        pass
+
+    @property
+    @abc.abstractmethod
+    def device(self) -> devices.Device:
+        pass
+
+    def freeze(self) -> 'cirq.FrozenCircuit':
+        '''Creates a FrozenCircuit from this circuit.
+
+        If 'self' is a FrozenCircuit, the original object is returned.
+        '''
+        from cirq.circuits import FrozenCircuit
+        if isinstance(self, FrozenCircuit):
+            return self
+        return FrozenCircuit(self,
+                             strategy=InsertStrategy.EARLIEST,
+                             device=self.device)
+
+    def unfreeze(self) -> 'cirq.Circuit':
+        '''Creates a Circuit from this circuit.
+
+        If 'self' is a Circuit, the original object is returned.
+        '''
+        if isinstance(self, Circuit):
+            return self
```

Comment:
There was an earlier [discussion](https://github.com/quantumlib/Cirq/pull/3445#discussion_r516197448) about a general protocol for "freezing" objects. The analogous method on `PauliString` is [`mutable_copy`](https://github.com/quantumlib/Cirq/blob/master/cirq/ops/pauli_string.py#L330) which makes it clear that a copy will be produced if needed, so that the resulting object can be modified as desired without affecting any existing objects. Without that guarantee I think `unfreeze` will be hard to use safely, so I'd suggest returning a copy here. We can think about a more generic protocol but I think that is out of scope for this PR.

Reply:
I was originally worried about having `unfreeze` return a copy while `freeze` returns the original, but given that `FrozenCircuit` is immutable I don't think there's any actual issue there. Will update tests to match.